File: Lua/Framework/UI/UIProgress.lua

UIProgressMode = {
	Horizontal = 1,
	Vertical = 2
}
UIProgress = {}

function UIProgress:New(go,progressMode)
	local o = {gameObject = go}
	o.onProgressChange = {}
	o.sprite = o.gameObject:GetComponent("UISprite")
	o.percent = -1
	o.progressMode = progressMode
	if(o.progressMode == UIProgressMode.Horizontal) then
		o.totalValue = o.sprite.width
	else
		o.totalValue = o.sprite.height
	end
	setmetatable(o,self)
	self.__index = self
	return o
end

function UIProgress:UpdateProgress(percent)
	if(self.sprite == nil) then return end
	if(percent < 0) then 
		percent = 0
	elseif(percent > 1) then
		percent = 1
	end
	if(percent ~= self.percent) then
		self.percent = percent
		self.sprite.gameObject:SetActive(self.percent ~= 0)
		local changeValue = self:GetPercent() * self.totalValue
		if(self.progressMode == UIProgressMode.Horizontal) then
			self.sprite.width = changeValue
		else
			self.sprite.height = changeValue
		end
		self:ChangeProcess(changeValue)
	end
end

function UIProgress:GetPercent()
	return self.percent
end

--添加进度改变事件
function UIProgress:AddProgressChangeFunc(func)
	if(not table.contains(self.onProgressChange,func)) then
		table.insert(self.onProgressChange,func)
	end
end

--移除进度改变事件
function UIProgress:RemoveProgressChangeFunc(func)
	local index = -1
	for i,v in ipairs(self.onProgressChange) do
		if(v == func) then
			index = i
		end
	end
	if(index > 0) then
		table.remove(self.onProgressChange,index)
	end
end

function UIProgress:ChangeProcess(value)
	if(self.onProgressChange ~= nil) then
		for k,v in pairs(self.onProgressChange) do
			if(v ~= nil) then
				v(value)
			end
		end
	end
end

function UIProgress:Dispose()
	self.onProgressChange = nil
	self.gameObject = nil
	self.sprite = nil
end

File: Script/Launch/View/UIProgress.cs
using System;
using UnityEngine;
using UnityEngine.UI;

namespace Launch
{
    public class UIProgress : MonoBehaviour
    {
        public event Action<float> OnProgressChange;
        protected UISprite _image;
        protected float _percent;
        protected float _totalValue;
        [SerializeField]
        private ProgressMode _progressMode = ProgressMode.Horizontal;

        private bool hasInit;
        protected virtual void Awake()
        {
            if (hasInit)
                return;
            hasInit = true;

            _image = gameObject.GetComponent<UISprite>();
            _percent = -1;
            SetProgressMode(_progressMode);
        }

        public virtual void Init()
        {
            Awake();
        }

        public void SetProgressMode(UIProgress.ProgressMode mode)
        {
            _progressMode = mode;
            if (_progressMode == ProgressMode.Horizontal)
            {
                _totalValue = _image.width;
            }
            else
            {
                _totalValue = _image.height;
            }
        }

        public void UpdateProgress(float percent)
        {
            if (_image == null) return;
            if (percent < 0) percent = 0f;
            else if (percent > 1) percent = 1f;

            if (_percent != percent)
            {
                _percent = percent;
                _image.gameObject.SetActive(_percent != 0);
                float changeValue = GetPercent() * _totalValue;
                
                if (_progressMode == ProgressMode.Horizontal)
                {
                    _image.width = (int)changeValue;
                }
                else
                {
                    _image.height = (int)changeValue;
                }
                ChangeProcess(_percent);
            }
        }

        public virtual float GetPercent()
        {
            return _percent;
        }

        protected void ChangeProcess(float value)
        {
            if (OnProgressChange != null)
            {
                OnProgressChange(value);
            }
        }

        protected bool IsFullProgress()
        {
            return _percent == 1f;
        }

        public enum ProgressMode
        {
            Horizontal,
            Vertical
        }
    }
}

